src/NServiceBus.RabbitMQ/MessagePump.cs

```diff
@@ -0,0 +1,163 @@
+﻿namespace NServiceBus.Transports.RabbitMQ
+{
+    using Extensibility;
+    using global::RabbitMQ.Client;
+    using global::RabbitMQ.Client.Events;
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Threading.Tasks;
+    using NServiceBus.Logging;
+    using NServiceBus.Transports.RabbitMQ.Config;
+    using NServiceBus.Transports.RabbitMQ.Connection;
+
+    class MessagePump : IPushMessages
+    {
+        readonly ReceiveOptions receiveOptions;
+        ConnectionConfiguration connectionConfiguration;
+        PoisonMessageForwarder poisonMessageForwarder;
+        QueuePurger queuePurger;
+
+        Func<PushContext, Task> pipe;
+        //CriticalError criticalError;
+        PushSettings settings;
+        SecondaryReceiveSettings secondaryReceiveSettings;
+
+        EventingBasicConsumer consumer;
+        TaskCompletionSource<bool> consumerShutdownCompleted;
+        private bool noAck;
+
+        public MessagePump(ReceiveOptions receiveOptions, ConnectionConfiguration connectionConfiguration, PoisonMessageForwarder poisonMessageForwarder, QueuePurger queuePurger)
+        {
+            this.receiveOptions = receiveOptions;
+            this.connectionConfiguration = connectionConfiguration;
+            this.poisonMessageForwarder = poisonMessageForwarder;
+            this.queuePurger = queuePurger;
+        }
+
+        public Task Init(Func<PushContext, Task> pipe, CriticalError criticalError, PushSettings settings)
+        {
+            this.pipe = pipe;
+            //this.criticalError = criticalError;
+            this.settings = settings;
+
+            secondaryReceiveSettings = receiveOptions.GetSettings(settings.InputQueue);
+            noAck = settings.RequiredTransactionMode == TransportTransactionMode.None;
+
+            if (settings.PurgeOnStartup)
+            {
+                queuePurger.Purge(settings.InputQueue);
+            }
+
+            return TaskEx.Completed;
+        }
+
+        public void Start(PushRuntimeSettings limitations)
+        {
+            var taskScheduler = new LimitedConcurrencyLevelTaskScheduler(limitations.MaxConcurrency);
+            var factory = new RabbitMqConnectionFactory(connectionConfiguration, taskScheduler);
+            var connection = new PersistentConnection(factory, connectionConfiguration.RetryDelay, "Consume");
+            var model = connection.CreateModel();
+            model.BasicQos(0, GetPrefetchGount(limitations.MaxConcurrency), false);
```

Comment:
Right, I was thinking the same thing, which is one of the reasons I hadn't added it back in yet. I was planning on doing some investigation on the ideal setting for that.